File: 0303.RangeSumQuery-Immutable.go
type NumArray struct {
    dp []int
}


func Constructor(nums []int) NumArray {
    sums := make([]int, len(nums)+1)
    for i := 0; i < len(nums); i++ {
        sums[i+1] = sums[i] + nums[i]
    }
    return NumArray{dp: sums}
}


func (this *NumArray) SumRange(i int, j int) int {
    return this.dp[j+1] - this.dp[i]
}


/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(i,j);
 */

File: 0000-0999/0303.RangeSumQuery-Immutable.py
class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        self.dp = nums
        for i in xrange(1, len(nums)):
            self.dp[i] += self.dp[i-1]

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        '''
        if i == 0:
            return self.dp[j]
        else:
            return self.dp[j] - self.dp[i-1]
        '''
        return self.dp[j] - (self.dp[i-1] if i else 0)


# Your NumArray object will be instantiated and called as such:
# numArray = NumArray(nums)
# numArray.sumRange(0, 1)
# numArray.sumRange(1, 2)